### PR TITLE
Carly and Kat edges adagrams team katillac project

### PR DESCRIPTION
**Pull request by kat-perreira**
# Adagrams
Congratulations! You're submitting your assignment.

## Comprehension Questions

Feature | Feedback
--- | ---
What are the components that make up a method? |  starts by defining the method signature, may have some parameters, then there is the method body and then returns only one thing, followed by an end of block
What are the advantages of using git when collaboratively working on one code base? | when working with git, we are able to save and share our files and share the changes made. Git also tracks our changes so we can go back to various versions of what we did.
What kind of relationship did you and your pair have with the unit tests? | we had a love hate relationship with the tests. We let it drive the direction of the program specifications and it allowed us to clarify and learn from our mistakes. But the ones that remained unpassed drove us batty.
Does your code use any methods from the `Enumerable` mixin? If so, where and why was it helpful? |nope, we could have used map and also sort_by.
What was one method you and your pair used to debug code? | we used binding.pry and the pry gem frequently to isolate variables to see what was happening and what was going wrong.
What are two discussion points that you and your pair discussed when giving/receiving feedback from each other that you would be willing to share? | If you're asking whether we discussed how we are pairing together, etc.  Yes, we discussed our methods of thinking individually and then sharing thoughts as a pair.  Then we would develop a plan (in psuedocode, drawing, etc.) before we started typing.  If you're asking what we learned, as a pair, from working together, we definitely discovered the benefits of pair programming - relying on one another's strengths and learning from one another.

### CODE REVIEW

**Comment by tildeee** [comment]
# Adagrams
## What We're Looking For

Feature | Feedback
--- | ---
**General** |
Answered comprehension questions | x
Both teammates contributed to the codebase | x
Small commits with meaningful commit messages | x
**Code Requirements** |
`draw_letters` method | x
Uses appropriate data structure to store the letter distribution | x
All tests for `draw_letters` pass | x
`uses_available_letters?` method | x, would probably go for a different approach
All tests for `uses_available_letters?` pass | x
`score_word` method | x, would probably approach it from a similar approach but different angle
Uses appropriate data structure to store the letter scores | x
All tests for `score_word` pass | x
`highest_score_from` method | x
Appropriately handles edge cases for tie-breaking logic | didn't quite get it!
All tests for `highest_score_from` pass |
**Overall** |

Good work y'all! Your code passes most of the tests

Overall I think that you all did a great job employing strategies to iterate through data structures and get the values you needed out of them. For example, I thought the way you all decided to make an array of hashes as `BANANA_BAG`, then made the method `making_tiles` to ultimately create a huge array of tiles was clever

That being said, I want to highlight two different things I noticed:

For the method `uses_available_letters?`, your approach was to iterate over `user_array` (which was the user input as an array of characters). Then, as you iterated, you asked if `letters_in_hand` included specifically the **last** character in `user_array`. If it did, then you deleted one character from `letters_in_hand` and the last character in `user_array` (using `reverse` and then `pop`). I would ask:

- What are the benefits of using the `.each` loop to iterate over every character?
- What are the benefits of looking at one character at a time with `user_array.last`, and then deleting/popping it one at a time?
- Do you need both?

For the method `score_word`, your approach was to iterate through every element in the `POINTS` data structure. When you looked at a hash in the `POINTS` structure, then you iterated through that, and after that, you iterated through every character in the user input word (`words_arr`). If the letter and the key were the same, you added the points.

I think that works, but if I were to rewrite that, I would look through each letter in the user input word (`words_arr`), and then try to access the points from there

Ultimately, I think you all definitely showed your skills about iterating over data and using them to your bidding. However, I think that if you look at your solutions with fresh eyes, you may be able to think of different angles of how to get to the same solution!


I have a few more comments that are simply about code style-- no big deals, just small stylistic things.

File: lib/adagrams.rb [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+    current_bag = letters[i] * values[i]
+    bag << current_bag.split("")
+  end
+
+  return bag
+
+end
+
+test = making_tiles(BANANA_BAG)
+BIG_BAG = test.flatten
```
Comment (tildeee) [inline lib/adagrams.rb:27]: normally, code that isn't located within a method won't run unless another part of the program calls it.

This code ends up executing when we run `rake` because our tests load this code and execute it... in general, I would expect these two lines to be the first lines in the `draw_letters` method actually

File: lib/adagrams.rb [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+      scores: score_word(word)
+    }
+    # shovelling the hashes above into the highest_scores array
+    highest_scores << score_hash
+    # shoveeling the return of the score_word method above (which is the score) for each of the words in the array
+    scores << score_word(word)
+  end
+
+  # sorts the scores from the scores array above and then makes the changes "permanent" with the ! so we can use it in the below iterations.
+  sorted_scores = scores.sort!
```
Comment (tildeee) [inline lib/adagrams.rb:135]: nice use of the `!`

File: lib/adagrams.rb [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+  winners_room = []
+  winner_hash = {}
+  # Iterating our way through the above array of hashes
+  highest_scores.each do |hash|
+    # Iterating through each of those hashes.  if the value of the score is equal to the highest member of the sorted_scores array (the largest), the word will then be shovelled into the winner array.
+
+    if hash[:scores] == sorted_scores.last
```
Comment (tildeee) [inline lib/adagrams.rb:142]: you end up getting all the scores, putting them in `scores`, sorting them, and then using `.last` on it in order to get the highest word score.

Maybe you could use `scores.max` instead!
